### DISCUSSION TITLE
Complete to matching common substring

### DISCUSSION THREAD

**Discussion by plk** [minad/vertico#448] [Q&A] [answered]
I am using vertico without orderless and want to get back the usual behaviour where tab completion fills the minibuffer with the common matching substring of matches. So, for example I have in the minibuffer:

```
/some/path/to/a/fi
```

and vertico is prompting with:

```
file1
file2
file3
```

I want TAB to make the minibuffer complete to:

```
/some/path/to/a/file
```

as with default emacs completion behaviour. Currently, TAB does nothing but display the vertico matches. Can't work out how to do this with vertico. Any help appreciated.

**Comment by minad** [comment]
See https://github.com/minad/vertico?tab=readme-ov-file#completion-styles-and-tab-completion. You probably want to bind TAB to `minibuffer-complete`.

**Reply by plk** [reply]
Already have:

```

(use-package! vertico
  :config
  (keymap-set vertico-map "TAB" #'minibuffer-complete)
  (keymap-set vertico-map "DEL" #'backward-delete-char)
  (setq read-file-name-completion-ignore-case nil)
  (setq read-buffer-completion-ignore-case nil)
  (setq completion-ignore-case nil)
  )
```

but TAB doesn't complete to the minimal matching substring.

For example, I'm pressing TAB in the screenshot but nothing happens. I want the minibuffer to complete to:

```
~/data/code/biblatex
```

<img width="520" alt="Screenshot 2024-02-23 at 22 09 16" src="https://github.com/minad/vertico/assets/1220851/88ea9e02-41ad-46e8-873d-25862439b854">

**Reply by minad** [reply]
Starting from emacs -Q the following works for me.

~~~ elisp
(package-initialize)
(vertico-mode 1)
(keymap-set vertico-map "TAB" #'minibuffer-complete)
~~~

Make sure that the `basic` or `substring` completion style is used.

**Reply by plk** [reply]
Hmm, `completion-styles` is

```
(basic partial-completion emacs22)
```

are you sure it works for you like I described? I tried with a bare config file with only the lines you describe and it's the same issue for me - TAB doesn't complete to the minimal shared substring.

**Reply by minad** [reply]
> are you sure it works for you like I described?

Yes, of course. The three lines from above evaluated in the scratch buffer of emacs -Q yields the desired result. bib is completed to biblatex when pressing TAB. Thanks for biber/biblatex btw!

**Reply by plk** [reply]
For me `emacs -Q` doesn't have vertico loaded or in the path so the second line just fails. I'll see if I can make a more minimal setup to test. Glad to see you are a `biblatex/biber` user ...

**Reply by plk** [reply]
This minimal setup with `emacs -Q` works and I see that the desired behaviour is fine with this:

```
(add-to-list 'load-path "~/.config/emacs/.local/straight/repos/vertico/")
(add-to-list 'load-path "~/.config/emacs/.local/straight/repos/compat/")
(require 'vertico)
(package-initialize)
(vertico-mode 1)
(keymap-set vertico-map "TAB" #'minibuffer-complete)
```

Now I have to try to find out what is stopping this in my config.

**Reply by minad** [reply]
It could be that you have some other completion styles configured via `completion-category-overrides` or `completion-category-defaults`. These will take precedence over the `completion-styles` variable.

**Reply by plk** [reply]
No customisation of those two - `completion-category-overrides` is nil and setting `completion-category-defaults` to nil doesn't help. I did notice that in the `emacs -Q` setup, `completion-styles` is undefined though ...

**Reply by minad** [reply]
It seems something is off with your setup or installation. By default `completion-styles` has the value `(basic partial-completion emacs22)`. This variable is preloaded as part of minibuffer.el and certainly not undefined.

**Comment by plk** [comment]
Found it. Consult mode was being used and so it had to be set like this:

```
(use-package! consult
  :config
  (consult-customize consult-completion-in-region
                     :completion-styles '(basic))
  )

```

**Reply by plk** [reply]
Spoke too soon. It seemed to work but then I restarted Emacs and it doesn't again. Grr. In fact this makes no difference and doesn't seem to be recognised. It falls back on base `completion-styles` and even when this is set to just `(basic)` continues to not work for me. It obviously fails somewhere in the guts of `consult--with-preview` but I can't tell where yet.

**Comment by plk** [comment, marked as answer]
Finally found the issue. Vertico uses `consult-completion-in-region` which was the issue. Setting:

```
  (setq completion-in-region-function 'completion--in-region)
```

Fixes it. It seems to go into a recursive minibuffer thing with the consult function and you need to press return as many times as you pressed TAB to complete to get the actual file selected to display.
